File: python/roteiro/indoor.py
"""Indoor GIS operations.

Provides functions for managing indoor buildings, floors, spaces, assets,
navigation, and importing IndoorGML/IFC models.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Dict, List, Optional

from .client import _encode_path_value, _with_query
from .models import (
    IndoorAsset,
    IndoorBuilding,
    IndoorFloor,
    IndoorModel,
    IndoorSpace,
    NavigationResult,
)

if TYPE_CHECKING:
    from .client import RoteiroClient


# ---------------------------------------------------------------------------
# Building CRUD
# ---------------------------------------------------------------------------


def list_buildings(client: RoteiroClient) -> List[IndoorBuilding]:
    """List all indoor buildings.

    Args:
        client: An initialised RoteiroClient instance.

    Returns:
        A list of IndoorBuilding objects.
    """
    data = client._get("/api/indoor/buildings")
    return [IndoorBuilding.from_dict(b) for b in data]


def get_building(client: RoteiroClient, building_id: str) -> IndoorBuilding:
    """Get a building with its floors and transitions.

    Args:
        client: An initialised RoteiroClient instance.
        building_id: The building identifier.

    Returns:
        An IndoorBuilding object with nested floors and transitions.
    """
    data = client._get(f"/api/indoor/buildings/{_encode_path_value(building_id)}")
    return IndoorBuilding.from_dict(data)


def create_building(client: RoteiroClient, data: Dict[str, Any]) -> IndoorBuilding:
    """Create a new building with optional floors, spaces, and transitions.

    Args:
        client: An initialised RoteiroClient instance.
        data: Building definition dict with ``id``, ``name``, and optionally
            ``address``, ``metadata``, ``floors``, and ``transitions``.

    Returns:
        The created IndoorBuilding.
    """
    resp = client._post("/api/indoor/buildings", data)
    return IndoorBuilding.from_dict(resp)


def update_building(
    client: RoteiroClient,
    building_id: str,
    data: Dict[str, Any],
) -> IndoorBuilding:
    """Update an existing building's metadata.

    Args:
        client: An initialised RoteiroClient instance.
        building_id: The building identifier.
        data: Fields to update (``name``, ``address``, ``metadata``, ``bounds``).

    Returns:
        The updated IndoorBuilding.
    """
    resp = client._put(
        f"/api/indoor/buildings/{_encode_path_value(building_id)}",
        data,
    )
    return IndoorBuilding.from_dict(resp)


def delete_building(client: RoteiroClient, building_id: str) -> None:
    """Delete a building and all its associated data.

    Args:
        client: An initialised RoteiroClient instance.
        building_id: The building identifier.
    """
    client._delete(f"/api/indoor/buildings/{_encode_path_value(building_id)}")


# ---------------------------------------------------------------------------
# Floor operations
# ---------------------------------------------------------------------------


def list_floors(client: RoteiroClient, building_id: str) -> List[IndoorFloor]:
    """List all floors in a building.

    Args:
        client: An initialised RoteiroClient instance.
        building_id: The building identifier.

    Returns:
        A list of IndoorFloor objects.
    """
    data = client._get(
        f"/api/indoor/buildings/{_encode_path_value(building_id)}/floors"
    )
    return [IndoorFloor.from_dict(f) for f in data]


def create_floor(
    client: RoteiroClient,
    building_id: str,
    floor_data: Dict[str, Any],
) -> IndoorFloor:
    """Create a new floor in a building.

    Args:
        client: An initialised RoteiroClient instance.
        building_id: The building identifier.
        floor_data: Floor definition with ``id``, ``name``, ``level``, etc.

    Returns:
        The created IndoorFloor.
    """
    resp = client._post(
        f"/api/indoor/buildings/{_encode_path_value(building_id)}/floors",
        floor_data,
    )
    return IndoorFloor.from_dict(resp)


# ---------------------------------------------------------------------------
# Space operations
# ---------------------------------------------------------------------------


def list_spaces(
    client: RoteiroClient,
    building_id: str,
    level: int,
) -> List[IndoorSpace]:
    """List all spaces on a given floor level.

    Args:
        client: An initialised RoteiroClient instance.
        building_id: The building identifier.
        level: The floor level number.

    Returns:
        A list of IndoorSpace objects.
    """
    data = client._get(
        f"/api/indoor/buildings/{_encode_path_value(building_id)}/floors/{_encode_path_value(level)}/spaces"
    )
    return [IndoorSpace.from_dict(s) for s in data]


def create_space(
    client: RoteiroClient,
    building_id: str,
    space_data: Dict[str, Any],
) -> IndoorSpace:
    """Create a new space in a building.

    Args:
        client: An initialised RoteiroClient instance.
        building_id: The building identifier.
        space_data: Space definition with ``id``, ``floor_id``, ``name``, etc.

    Returns:
        The created IndoorSpace.
    """
    resp = client._post(
        f"/api/indoor/buildings/{_encode_path_value(building_id)}/spaces",
        space_data,
    )
    return IndoorSpace.from_dict(resp)


def get_space(
    client: RoteiroClient,
    building_id: str,
    space_id: str,
) -> IndoorSpace:
    """Get a specific space within a building.

    Args:
        client: An initialised RoteiroClient instance.
        building_id: The building identifier.
        space_id: The space identifier.

    Returns:
        An IndoorSpace object.
    """
    data = client._get(
        f"/api/indoor/buildings/{_encode_path_value(building_id)}/spaces/{_encode_path_value(space_id)}"
    )
    return IndoorSpace.from_dict(data)


# ---------------------------------------------------------------------------
# Asset operations
# ---------------------------------------------------------------------------


def list_assets(
    client: RoteiroClient,
    building_id: str,
    floor_id: Optional[str] = None,
    space_id: Optional[str] = None,
) -> List[IndoorAsset]:
    """List assets in a building, optionally filtered by floor or space.

    Args:
        client: An initialised RoteiroClient instance.
        building_id: The building identifier.
        floor_id: Optional floor ID to filter by.
        space_id: Optional space ID to filter by.

    Returns:
        A list of IndoorAsset objects.
    """
    path = _with_query(
        f"/api/indoor/buildings/{_encode_path_value(building_id)}/assets",
        [("floor", floor_id), ("space", space_id)],
    )
    data = client._get(path)
    return [IndoorAsset.from_dict(a) for a in data]


def create_asset(
    client: RoteiroClient,
    building_id: str,
    asset_data: Dict[str, Any],
) -> IndoorAsset:
    """Create a new asset in a building.

    Args:
        client: An initialised RoteiroClient instance.
        building_id: The building identifier.
        asset_data: Asset definition with ``id``, ``name``, ``floor_id``,
            ``space_id``, etc.

    Returns:
        The created IndoorAsset.
    """
    resp = client._post(
        f"/api/indoor/buildings/{_encode_path_value(building_id)}/assets",
        asset_data,
    )
    return IndoorAsset.from_dict(resp)


# ---------------------------------------------------------------------------
# Navigation
# ---------------------------------------------------------------------------


def find_path(
    client: RoteiroClient,
    building_id: str,
    from_space: str,
    to_space: str,
    accessible_only: bool = False,
) -> NavigationResult:
    """Find an indoor navigation path between two spaces.

    Args:
        client: An initialised RoteiroClient instance.
        building_id: The building identifier.
        from_space: Source space ID.
        to_space: Destination space ID.
        accessible_only: If True, only use accessible transitions.

    Returns:
        A NavigationResult with path steps, total distance, and time estimate.
    """
    data = client._post(
        "/api/indoor/navigate",
        {
            "building_id": building_id,
            "from": from_space,
            "to": to_space,
            "accessible_only": accessible_only,
        },
    )
    return NavigationResult.from_dict(data)


def find_nearest(
    client: RoteiroClient,
    building_id: str,
    from_space: str,
    target_type: str,
    accessible_only: bool = False,
) -> NavigationResult:
    """Find the nearest space of a given type from a starting space.

    Args:
        client: An initialised RoteiroClient instance.
        building_id: The building identifier.
        from_space: Source space ID.
        target_type: Type of space to search for (e.g. ``restroom``, ``exit``).
        accessible_only: If True, only use accessible transitions.

    Returns:
        A NavigationResult with path steps, total distance, and time estimate.
    """
    data = client._post(
        "/api/indoor/navigate/nearest",
        {
            "building_id": building_id,
            "from": from_space,
            "target_type": target_type,
            "accessible_only": accessible_only,
        },
    )
    return NavigationResult.from_dict(data)


# ---------------------------------------------------------------------------
# Import
# ---------------------------------------------------------------------------


def parse_indoor_gml(client: RoteiroClient, file_path: str) -> IndoorModel:
    """Import and parse an IndoorGML file.

    Uploads the file to the server for parsing and creates the building
    model in the indoor GIS system.

    Args:
        client: An initialised RoteiroClient instance.
        file_path: Local path to the IndoorGML (.gml/.xml) file.

    Returns:
        An IndoorModel containing the parsed building.
    """
    data = client._upload_file("/api/indoor/import", file_path)
    return IndoorModel.from_dict(data)


def import_ifc(client: RoteiroClient, file_path: str) -> IndoorModel:
    """Import and parse an IFC (Industry Foundation Classes) file.

    Uploads the file to the server for parsing and creates the building
    model in the indoor GIS system.

    Args:
        client: An initialised RoteiroClient instance.
        file_path: Local path to the IFC file.

    Returns:
        An IndoorModel containing the parsed building.
    """
    data = client._upload_file("/api/indoor/import", file_path)
    return IndoorModel.from_dict(data)


# ---------------------------------------------------------------------------
# Occupancy & Analytics
# ---------------------------------------------------------------------------


def get_occupancy(client: RoteiroClient, building_id: str) -> Dict[str, Any]:
    """Get occupancy analytics for a building.

    Args:
        client: An initialised RoteiroClient instance.
        building_id: The building identifier.

    Returns:
        A dictionary with occupancy data per floor and space.
    """
    return client._get(
        f"/api/indoor/buildings/{_encode_path_value(building_id)}/occupancy"
    )


def get_evacuation_routes(
    client: RoteiroClient,
    building_id: str,
    params: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """Compute evacuation routes for a building.

    Args:
        client: An initialised RoteiroClient instance.
        building_id: The building identifier.
        params: Optional parameters for evacuation routing.

    Returns:
        A dictionary with evacuation route data.
    """
    return client._post(
        f"/api/indoor/buildings/{_encode_path_value(building_id)}/evacuation",
        params or {},
    )

File: typescript/src/indoor.ts
/**
 * Indoor GIS operations.
 *
 * Provides functions for managing buildings, floors, spaces, assets,
 * navigation, and importing IndoorGML/IFC models.
 * @module indoor
 */

import type { RoteiroClient } from './client';
import type {
  IndoorAsset,
  IndoorBuilding,
  IndoorFloor,
  IndoorModel,
  IndoorSpace,
  NavigationResult,
  OccupancyData,
} from './types';

// ---------------------------------------------------------------------------
// Buildings
// ---------------------------------------------------------------------------

/**
 * List all indoor buildings.
 *
 * @param client - An initialised RoteiroClient instance.
 * @returns A list of IndoorBuilding objects.
 */
export async function listBuildings(
  client: RoteiroClient,
): Promise<IndoorBuilding[]> {
  return client.request<IndoorBuilding[]>('/api/indoor/buildings');
}

/**
 * Get a building with its floors and transitions.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param buildingId - The building identifier.
 * @returns An IndoorBuilding with nested floors and transitions.
 */
export async function getBuilding(
  client: RoteiroClient,
  buildingId: string,
): Promise<IndoorBuilding> {
  return client.request<IndoorBuilding>(
    `/api/indoor/buildings/${encodeURIComponent(buildingId)}`,
  );
}

/**
 * Create a new building with optional floors, spaces, and transitions.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param data - Building definition with id, name, floors, transitions, etc.
 * @returns The created IndoorBuilding.
 */
export async function createBuilding(
  client: RoteiroClient,
  data: {
    id: string;
    name: string;
    address?: string;
    metadata?: Record<string, string>;
    floors?: Array<{
      id: string;
      name: string;
      level: number;
      elevation?: number;
      height?: number;
      spaces?: Array<{
        id: string;
        name: string;
        space_type?: string;
        geometry?: unknown;
        navigable?: boolean;
        connections?: string[];
      }>;
    }>;
    transitions?: Array<{
      id: string;
      name: string;
      transition_type?: string;
      connects_floors: number[];
      accessible?: boolean;
      bidirectional?: boolean;
      space_ids?: string[];
    }>;
  },
): Promise<IndoorBuilding> {
  return client.post<IndoorBuilding>('/api/indoor/buildings', data);
}

/**
 * Update an existing building's metadata.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param buildingId - The building identifier.
 * @param data - Fields to update (name, address, metadata, bounds).
 * @returns The updated IndoorBuilding.
 */
export async function updateBuilding(
  client: RoteiroClient,
  buildingId: string,
  data: Partial<IndoorBuilding>,
): Promise<IndoorBuilding> {
  return client.put<IndoorBuilding>(
    `/api/indoor/buildings/${encodeURIComponent(buildingId)}`,
    data,
  );
}

/**
 * Delete a building and all its associated data.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param buildingId - The building identifier.
 */
export async function deleteBuilding(
  client: RoteiroClient,
  buildingId: string,
): Promise<void> {
  await client.del(`/api/indoor/buildings/${encodeURIComponent(buildingId)}`);
}

// ---------------------------------------------------------------------------
// Floors
// ---------------------------------------------------------------------------

/**
 * List all floors in a building.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param buildingId - The building identifier.
 * @returns A list of IndoorFloor objects.
 */
export async function listFloors(
  client: RoteiroClient,
  buildingId: string,
): Promise<IndoorFloor[]> {
  return client.request<IndoorFloor[]>(
    `/api/indoor/buildings/${encodeURIComponent(buildingId)}/floors`,
  );
}

/**
 * Create a new floor in a building.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param buildingId - The building identifier.
 * @param floorData - Floor definition with id, name, level, etc.
 * @returns The created IndoorFloor.
 */
export async function createFloor(
  client: RoteiroClient,
  buildingId: string,
  floorData: Partial<IndoorFloor>,
): Promise<IndoorFloor> {
  return client.post<IndoorFloor>(
    `/api/indoor/buildings/${encodeURIComponent(buildingId)}/floors`,
    floorData,
  );
}

// ---------------------------------------------------------------------------
// Spaces
// ---------------------------------------------------------------------------

/**
 * List all spaces on a given floor level.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param buildingId - The building identifier.
 * @param level - The floor level number.
 * @returns A list of IndoorSpace objects.
 */
export async function listSpaces(
  client: RoteiroClient,
  buildingId: string,
  level: number,
): Promise<IndoorSpace[]> {
  return client.request<IndoorSpace[]>(
    `/api/indoor/buildings/${encodeURIComponent(buildingId)}/floors/${level}/spaces`,
  );
}

/**
 * Create a new space in a building.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param buildingId - The building identifier.
 * @param spaceData - Space definition with id, floor_id, name, etc.
 * @returns The created IndoorSpace.
 */
export async function createSpace(
  client: RoteiroClient,
  buildingId: string,
  spaceData: Partial<IndoorSpace>,
): Promise<IndoorSpace> {
  return client.post<IndoorSpace>(
    `/api/indoor/buildings/${encodeURIComponent(buildingId)}/spaces`,
    spaceData,
  );
}

/**
 * Get a specific space within a building.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param buildingId - The building identifier.
 * @param spaceId - The space identifier.
 * @returns An IndoorSpace object.
 */
export async function getSpace(
  client: RoteiroClient,
  buildingId: string,
  spaceId: string,
): Promise<IndoorSpace> {
  return client.request<IndoorSpace>(
    `/api/indoor/buildings/${encodeURIComponent(buildingId)}/spaces/${encodeURIComponent(spaceId)}`,
  );
}

// ---------------------------------------------------------------------------
// Assets
// ---------------------------------------------------------------------------

/**
 * List assets in a building, optionally filtered by floor or space.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param buildingId - The building identifier.
 * @param floorId - Optional floor ID filter.
 * @param spaceId - Optional space ID filter.
 * @returns A list of IndoorAsset objects.
 */
export async function listAssets(
  client: RoteiroClient,
  buildingId: string,
  floorId?: string,
  spaceId?: string,
): Promise<IndoorAsset[]> {
  const params = new URLSearchParams();
  if (floorId) params.set('floor', floorId);
  if (spaceId) params.set('space', spaceId);
  const q = params.toString();
  return client.request<IndoorAsset[]>(
    `/api/indoor/buildings/${encodeURIComponent(buildingId)}/assets${q ? `?${q}` : ''}`,
  );
}

/**
 * Create a new asset in a building.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param buildingId - The building identifier.
 * @param assetData - Asset definition with id, name, floor_id, space_id, etc.
 * @returns The created IndoorAsset.
 */
export async function createAsset(
  client: RoteiroClient,
  buildingId: string,
  assetData: Partial<IndoorAsset>,
): Promise<IndoorAsset> {
  return client.post<IndoorAsset>(
    `/api/indoor/buildings/${encodeURIComponent(buildingId)}/assets`,
    assetData,
  );
}

// ---------------------------------------------------------------------------
// Navigation
// ---------------------------------------------------------------------------

/**
 * Find an indoor navigation path between two spaces.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param buildingId - The building identifier.
 * @param fromSpace - Source space ID.
 * @param toSpace - Destination space ID.
 * @param accessibleOnly - Only use accessible transitions (default: false).
 * @returns A NavigationResult with path steps, distance, and time estimate.
 */
export async function findPath(
  client: RoteiroClient,
  buildingId: string,
  fromSpace: string,
  toSpace: string,
  accessibleOnly: boolean = false,
): Promise<NavigationResult> {
  return client.post<NavigationResult>('/api/indoor/navigate', {
    building_id: buildingId,
    from: fromSpace,
    to: toSpace,
    accessible_only: accessibleOnly,
  });
}

/**
 * Find the nearest space of a given type from a starting space.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param buildingId - The building identifier.
 * @param fromSpace - Source space ID.
 * @param targetType - Type of space to search for (e.g. "restroom", "exit").
 * @param accessibleOnly - Only use accessible transitions (default: false).
 * @returns A NavigationResult with path steps, distance, and time estimate.
 */
export async function findNearest(
  client: RoteiroClient,
  buildingId: string,
  fromSpace: string,
  targetType: string,
  accessibleOnly: boolean = false,
): Promise<NavigationResult> {
  return client.post<NavigationResult>('/api/indoor/navigate/nearest', {
    building_id: buildingId,
    from: fromSpace,
    target_type: targetType,
    accessible_only: accessibleOnly,
  });
}

// ---------------------------------------------------------------------------
// Import
// ---------------------------------------------------------------------------

/**
 * Import an IndoorGML file to create a building model.
 *
 * Note: For file uploads, use a FormData approach or pass the file content
 * as the request body with the appropriate Content-Type header.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param fileContent - The IndoorGML file content as a string.
 * @param filename - Optional filename hint for format detection.
 * @returns The imported IndoorModel.
 */
export async function parseIndoorGml(
  client: RoteiroClient,
  fileContent: string,
  filename: string = 'model.gml',
): Promise<IndoorModel> {
  const resp = await client.request<IndoorBuilding>(
    `/api/indoor/import?filename=${encodeURIComponent(filename)}`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/xml' },
      body: fileContent,
    },
  );
  return { building: resp, source_format: 'indoorgml' };
}

/**
 * Import an IFC file to create a building model.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param fileContent - The IFC file content as a string.
 * @param filename - Optional filename hint for format detection.
 * @returns The imported IndoorModel.
 */
export async function importIfc(
  client: RoteiroClient,
  fileContent: string,
  filename: string = 'model.ifc',
): Promise<IndoorModel> {
  const resp = await client.request<IndoorBuilding>(
    `/api/indoor/import?filename=${encodeURIComponent(filename)}`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/octet-stream' },
      body: fileContent,
    },
  );
  return { building: resp, source_format: 'ifc' };
}

// ---------------------------------------------------------------------------
// Occupancy & Evacuation
// ---------------------------------------------------------------------------

/**
 * Get occupancy analytics for a building.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param buildingId - The building identifier.
 * @returns Occupancy data with per-floor and per-space breakdown.
 */
export async function getOccupancy(
  client: RoteiroClient,
  buildingId: string,
): Promise<OccupancyData> {
  return client.request<OccupancyData>(
    `/api/indoor/buildings/${encodeURIComponent(buildingId)}/occupancy`,
  );
}

/**
 * Compute evacuation routes for a building.
 *
 * @param client - An initialised RoteiroClient instance.
 * @param buildingId - The building identifier.
 * @param params - Optional evacuation routing parameters.
 * @returns Evacuation route data.
 */
export async function getEvacuationRoutes(
  client: RoteiroClient,
  buildingId: string,
  params?: Record<string, unknown>,
): Promise<unknown> {
  return client.post<unknown>(
    `/api/indoor/buildings/${encodeURIComponent(buildingId)}/evacuation`,
    params ?? {},
  );
}
